### PR TITLE
MQTT Hygiene: Remove "callback" label from getTime function

### PR DESCRIPTION
**Pull request by aggarw13**
**Purpose of PR**
Update MQTT library to not label the `getTime` member of `MQTTContext_t` as an "**ApplicationCallback**". `getTime` is not a callback but a utility functor required for the functioning of the library.

**Changes**
* Remove the `MQTTApplicationCallbacks_t` struct and keep the `getTime` and `appCallback` as separate members of `MQTTContext_t`
* Update `MQTT_Init` function to take the `getTime` and `appCallback` functors as separate input parameters
* Add the new parameters names of `MQTT_Init` to mqtt/lexicon.txt
* Fix `README` instructions for spell-checker's addition to system `PATH`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tools/spell/README.md [hunk 1]
```diff
@@ -1,15 +1,13 @@
 # How to create a lexicon.txt for a new library.
 
-1. In your GNU environment install programs: *spell* and *getopt*  
-   For Linux here are the commands to install these programs:  
+1. In your GNU environment install the *spell* program. Use the following command in Linux, to install the package:
    ```shell
```
Comment (aggarw13) [inline tools/spell/README.md:4]: `getopt` was already part of my default Ubunutu 18.04 installation. 
The command `apt-get install getopt` is incorrect in finding the package

File: libraries/standard/mqtt/integration-test/mqtt_system_test.c [hunk 2]
```diff
@@ -311,18 +312,19 @@ static void establishMqttSession( MQTTContext_t * pContext,
...
+    /* User-callback for receiving incoming publishes and incoming acks
      * from MQTT library. */
-    callbacks.appCallback = eventCallback;
+    userCallback = eventCallback;
 
-    /* Application callback for getting the time for MQTT library. This time
+    /* Platform-specific utility function for getting the time for MQTT library. This time
      * function will be used to calculate intervals in MQTT library.*/
-    callbacks.getTime = Clock_GetTimeMs;
+    getTimeFunction = Clock_GetTimeMs;
```
Comment (leegeth) [inline libraries/standard/mqtt/integration-test/mqtt_system_test.c:321]: In demos, these additional steps are providing more readability for the example application. Can we avoid it here and directly pass the function pointer in MQTT_Init?

File: libraries/standard/mqtt/utest/mqtt_utest.c [hunk 3]
```diff
@@ -554,42 +541,33 @@ void test_MQTT_Init_Invalid_Params( void )
...
     TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
 
-    mqttStatus = MQTT_Init( &context, &transport, &callbacks, NULL );
+    mqttStatus = MQTT_Init( &context, &transport, getTime, eventCallback, NULL );
     TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
 
-    /* Test if NULL is passed for any callbacks. */
-    callbacks.getTime = NULL;
-    mqttStatus = MQTT_Init( &context, &transport, &callbacks, &networkBuffer );
+    /* Test if NULL is passed for any functors. */
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_utest.c:557]: I think we are just using function pointers here. Functors are slightly different from function pointers https://en.wikipedia.org/wiki/Function_pointer#Functors

File: tools/spell/README.md (hunk 1 above)
Comment (leegeth) [reply tools/spell/README.md:4]: Is this true about different linux distributions?

File: libraries/standard/mqtt/utest/mqtt_utest.c (hunk 3 above)
Comment (aggarw13) [reply libraries/standard/mqtt/utest/mqtt_utest.c:557]: Oops, me coming from C++ land, am used to calling it functors (where I mostly used `std::function` objects 😛)  Will remove use of functor

File: libraries/standard/mqtt/integration-test/mqtt_system_test.c (hunk 2 above)
Comment (aggarw13) [reply libraries/standard/mqtt/integration-test/mqtt_system_test.c]: Removed it

File: tools/spell/README.md (hunk 1 above)
Comment (aggarw13) [reply tools/spell/README.md:4]: Not sure about other distros. Will need to check it. However, `apt-get install getopt` command is incorrect...

Comment (abhidixi11) [reply tools/spell/README.md:4]: Debian family uses apt-get and Redhat family uses Yum.

Comment (leegeth) [reply tools/spell/README.md:4]: Yea. If it is incorrect, can we just remove it?Can we assume that the target system is Ubunutu 18.04?

Comment (aggarw13) [reply tools/spell/README.md:4]: Found from this manual (http://manpages.ubuntu.com/manpages/trusty/man1/getopt.1.html) that `util-linux` is the kernel package from which **getopt** can be installed. (Should be part of default installation in Linux environments)

Have updated to provide the correct command for installation of the utility.

File: libraries/standard/mqtt/src/mqtt.c [hunk 4]
```diff
@@ -1769,25 +1770,23 @@ MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * pContext,
...
     else if( pContext->networkBuffer.pBuffer == NULL )
     {
-        LogError( ( "The MQTT context's networkBuffer must not be NULL." ) );
+        LogError( ( "Invalid input parameter: The MQTT context's networkBuffer must not be NULL." ) );
     }
     else
     {
-        getTimeStampMs = pContext->callbacks.getTime;
-        entryTimeMs = getTimeStampMs();
+        entryTimeMs = pContext->getTime();
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt.c:1789]: I think it's odd you changed this for the process loop but not the receive loop. Why not keep them similar?

File: libraries/standard/mqtt/src/mqtt.c [hunk 5]
```diff
@@ -1832,19 +1832,19 @@ MQTTStatus_t MQTT_ReceiveLoop( MQTTContext_t * pContext,
...
     if( pContext == NULL )
     {
-        LogError( ( "MQTT Context cannot be NULL." ) );
+        LogError( ( "Invalid input parameter: MQTT Context cannot be NULL." ) );
     }
-    else if( pContext->callbacks.getTime == NULL )
+    else if( pContext->getTime == NULL )
     {
-        LogError( ( "MQTT Context must set callbacks.getTime." ) );
+        LogError( ( "Invalid input parameter: MQTT Context have valid getTime function." ) );
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt.c:1839]: Grammar
```suggestion
        LogError( ( "Invalid input parameter: MQTT Context must have a valid getTime function." ) );
```

File: libraries/standard/mqtt/src/mqtt.c (hunk 4 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/mqtt.c:1789]: Thanks for catching. Made the two consistent

File: demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c [hunk 6]
```diff
@@ -660,7 +660,8 @@ static int establishMqttSession( MQTTContext_t * pMqttContext,
     MQTTStatus_t mqttStatus;
     MQTTConnectInfo_t connectInfo;
     MQTTFixedBuffer_t networkBuffer;
-    MQTTApplicationCallbacks_t callbacks;
+    MQTTEventCallback_t userCallback;
```
Comment (aggarg) [inline demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c:663]: Are these two local variables needed anymore? Can we not directly have:

```c
MQTT_Init( pMqttContext, &transport, Clock_GetTimeMs, eventCallback, &networkBuffer );
```

File: demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c [hunk 7]
```diff
@@ -706,16 +707,16 @@ static int establishMqttSession( MQTTContext_t * pMqttContext,
...
 
-    /* Application callback for getting the time for MQTT library. This time
+    /* Platform-specific utility function for getting the time for MQTT library. This time
      * function will be used to calculate intervals in MQTT library.*/
-    callbacks.getTime = Clock_GetTimeMs;
+    getTimeFunction = Clock_GetTimeMs;
 
     /* Initialize MQTT library. */
-    mqttStatus = MQTT_Init( pMqttContext, &transport, &callbacks, &networkBuffer );
+    mqttStatus = MQTT_Init( pMqttContext, &transport, getTimeFunction, userCallback, &networkBuffer );
```
Comment (aggarg) [inline demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:719]: Same as above.

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 8]
```diff
@@ -443,16 +444,16 @@ static int establishMqttSession( MQTTContext_t * pMqttContext,
...
 
-    /* Application callback for getting the time for MQTT library. This time
+    /* Platform-specific utility function for getting the time for MQTT library. This time
      * function will be used to calculate intervals in MQTT library.*/
-    callbacks.getTime = Clock_GetTimeMs;
+    getTimeFunction = Clock_GetTimeMs;
 
     /* Initialize MQTT library. */
-    mqttStatus = MQTT_Init( pMqttContext, &transport, &callbacks, &networkBuffer );
+    mqttStatus = MQTT_Init( pMqttContext, &transport, getTimeFunction, userCallback, &networkBuffer );
```
Comment (aggarg) [inline demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:456]: Same as above.

File: libraries/standard/mqtt/include/mqtt.h [hunk 9]
```diff
@@ -143,15 +138,18 @@ struct MQTTContext
...
+ * incoming packet events.
  * @brief param[in] pNetworkBuffer Network buffer provided for the context.
  *
  * @return #MQTTBadParameter if invalid parameters are passed;
  * #MQTTSuccess otherwise.
  */
 MQTTStatus_t MQTT_Init( MQTTContext_t * pContext,
                         const TransportInterface_t * pTransportInterface,
-                        const MQTTApplicationCallbacks_t * pCallbacks,
+                        const MQTTGetCurrentTimeFunc_t getTimeFunction,
```
Comment (aggarg) [inline libraries/standard/mqtt/include/mqtt.h:151]: These are no longer pointers, remove const.

File: libraries/standard/mqtt/src/mqtt.c [hunk 10]
```diff
@@ -1313,31 +1313,31 @@ static MQTTStatus_t validatePublishParams( const MQTTContext_t * pContext,
 
 MQTTStatus_t MQTT_Init( MQTTContext_t * pContext,
                         const TransportInterface_t * pTransportInterface,
-                        const MQTTApplicationCallbacks_t * pCallbacks,
+                        const MQTTGetCurrentTimeFunc_t getTimeFunction,
```
Comment (aggarg) [inline libraries/standard/mqtt/src/mqtt.c:1316]: Remove const.

File: demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c (hunk 6 above)
Comment (aggarw13) [reply demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c:663]: I added the local variables only for keeping the verbose documentation
```
  /* Application callbacks for receiving incoming publishes and incoming acks
     * from MQTT library. */
    callbacks.appCallback = eventCallback;

    /* Application callback for getting the time for MQTT library. This time
     * function will be used to calculate intervals in MQTT library.*/
    callbacks.getTime = Clock_GetTimeMs;
```

File: demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c (hunk 7 above)
Comment (aggarw13) [reply demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c]: Removed

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c (hunk 8 above)
Comment (aggarw13) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c]: Removed

File: libraries/standard/mqtt/src/mqtt.c (hunk 10 above)
Comment (aggarw13) [reply libraries/standard/mqtt/src/mqtt.c]: Done

**Pull request merged**